### PR TITLE
Write secure HTTP API spike

### LINKED ISSUE #477: How to secure the `cnd` HTTP API with HTTPS?
Answer the following questions:

- comit-i should be able to be served via HTTPS
- API should be accessible via HTTPS
- how to install the certificate?
- does it have to be self-signed? is there any funny workaround for localhost-only HTTPS APIs?

### PR DESCRIPTION
**Pull request by tcharding**
As defined in the issue [below], we would like to consider securing the HTTP API exposed by cnd.  'securing' is an ambiguous term, this spike is limited in scope to encrypting the traffic between cnd and comit-i.

Fixes: https://github.com/comit-network/comit-rs/issues/477

### CODE REVIEW

**Review by bonomat — CHANGES_REQUESTED** [review]
Thanks for the summary. 

I think you can go one step further and propose a solution to how we can secure cnd's API exactly. I'd say *not* doing this is not an option if we want to write *best practice software* as you state at the beginning.

**Comment by tcharding** [comment]
> I'd say _not_ doing this is not an option if we want to write _best practice software_ as you state at the beginning.

Fair point, will do further research.

**Comment by tcharding** [comment]
Ok, spike new and revised - I'm happy with my conclusions and that someone could now go ahead and implement TLS as recommended in this spike without having to do further research.  Please say if this view is incorrect - I'd really like to get this whole spike writing thing under my belt.

**Review by thomaseizinger — COMMENTED** [review]
The problem with self-signed certs is that the browser will not trust them and display a giant warning to the user that what he tries to do is unsafe. IMO, this is not an acceptable solution.

If we want to use a self-signed certificate, I think we would need to install it as a CA on the computer (root privileges required) which seems to be bad.

This documentation here on letsencrypt.org goes into more detail for this problem: https://letsencrypt.org/docs/certificates-for-localhost/

Regarding the spike: I think the approach is good but don't think it will be possible to have a good solution to this if `comit-it` is out-of-scope :)

File: 0013-secure-http-api.adoc [hunk 1]
```diff
@@ -0,0 +1,85 @@
+= Securing the `cnd` HTTP API
+
```
Comment (D4nte) [inline 0013-secure-http-api.adoc]:
```suggestion
```
Must not have a new line for ascii doc to pick up the author line.

File: 0013-secure-http-api.adoc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+Tobin C. Harding <tobin.harding@coblox.tech>
+:toc:
+:revdate: 2019-08-13
+
+* Authors(s): {authors}
```
Comment (D4nte) [inline 0013-secure-http-api.adoc]:
```suggestion
* Authors(s): {authors} +
* Date: {revdate} +
```
There is a '+' in the template, not sure what was the reason but pretty sure you need to keep it

**Review by D4nte — COMMENTED** [review]
@tcharding can you do a pass on the asciidoc format? Just check the file from GitHub and correct what does not render well (new line is needed before a list for example).

File: 0013-secure-http-api.adoc (hunk 2 above)
Comment (thomaseizinger) [reply 0013-secure-http-api.adoc]: Is needed for correct rendering.

**Comment by tcharding** [comment]
> @tcharding can you do a pass on the asciidoc format? Just check the file from GitHub and correct what does not render well (new line is needed before a list for example).

My bad, thanks for poking me.

**Comment by tcharding** [comment]
> The problem with self-signed certs is that the browser will not trust them and display a giant warning to the user that what he tries to do is unsafe. IMO, this is not an acceptable solution.

Yes, I did not include that in the spike I assumed it was common knowledge and acceptable.

> If we want to use a self-signed certificate, I think we would need to install it as a CA on the computer (root privileges required) which seems to be bad.

This is worse IMO than just expecting the user to click through the warning - since our target are devs I feel it is safe to assume they know what doing that means and since they most likely started the cnd themselves this will not be a surprise.  Should this discussion go into the spike? 

> This documentation here on letsencrypt.org goes into more detail for this problem: [letsencrypt.org/docs/certificates-for-localhost](https://letsencrypt.org/docs/certificates-for-localhost/)
> 
> Regarding the spike: I think the approach is good but don't think it will be possible to have a good solution to this if `comit-it` is out-of-scope :)

Oh, does it read like `comit-i` is out of scope.  That was not my intention, I was hoping someone with front end knowledge would speak up.  Perhaps a better method of getting this information would be for me to actively ask :)

File: 0013-secure-http-api.adoc (hunk 2 above)
Comment (bonomat) [reply 0013-secure-http-api.adoc]: It's to get a new line after `* Authors(s): {authors}`
https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#breaks

File: 0013-secure-http-api.adoc [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+The only real consideration is the TLS keypair (key and certificate) and selfsigning the certificate.
+We can provide a default keypair so that cnd works out of the box with TLS.
+This can be done in a few ways:
+* Put the keys in the repository - bad, private keys should be private.
+* Write a standalone tool to generate a new key.
+* Add code to cnd to generate a new keypair if one is not found i.e. in the default place or via a config option `--key=/path/to/key`and `--cert=/path/to/cert`.
+
+=== comit-i
+
+Request content for this section to be written by team member who is familiar with `comit-i` and/or React.
```
Comment (bonomat) [inline 0013-secure-http-api.adoc]: In order to get some knowledge around comit-i, please pair up with someone and write it together.

**Comment by D4nte** [comment]
> This is worse IMO than just expecting the user to click through the warning - since our target are devs I feel it is safe to assume they know what doing that means and since they most likely started the cnd themselves this will not be a surprise. Should this discussion go into the spike?

Our target is devs that build apps used by users. In this case, the final user would still need to click through the warning if the app is the webapp which is not good.

_Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow_

File: 0013-secure-http-api.adoc (hunk 3 above)
Comment (tcharding) [reply 0013-secure-http-api.adoc]: > In order to get some knowledge around comit-i, please pair up with someone and write it together.

Ok

**Comment by tcharding** [comment]
> Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow

Where should these discussions be please @D4nte?  On the original issue?

**Comment by thomaseizinger** [comment]
> Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow

Probably my bad because I voiced the concern in the general review, with does not allow to create threads.

@tcharding 

Thanks for amending the spike with those concerns! How are we dealing with that problem now?

Didn't read it yet but maybe this HN thread has some insights: https://news.ycombinator.com/item?id=17358339

**Comment by tcharding** [comment]
> Didn't read it yet but maybe this HN thread has some insights: news.ycombinator.com/item?id=17358339

Good link, thanks.  According to the blog post linked (indirectly) above; we may be able to get around using self signed certificates by using `127.0.0.1` instead of `localhost`.  Turns out modern browsers treat these addresses differently and the IP version is considered ["potentially trustworthy"](https://www.w3.org/TR/secure-contexts/#is-origin-trustworthy).

**Comment by D4nte** [comment]
> > Also, please avoid starting discussions on the main thread for spikes, makes it hard to follow
> 
> Where should these discussions be please @D4nte? On the original issue?

In the "Files changed" tab, on the line concerned by the comment.

File: 0013-secure-http-api.adoc [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+.What TLS gives us and whether we need it:
+* Authentication of cnd - needed, allows client to verify it is connecting to the correct cnd.
+* Data confidentiality - needed, protects against attacker learning the secret.
+* Date integrity - needed, protects against modification of the data i.e. changing the redeem address.
+* Replay protection - not needed, replay attack vector currently non-existent for cnd.
+* Sequence protection - no obvious need.
+
+=== Certificates
+
+TLS requires a certificate (private/public keypair) to be preset on the server i.e., cnd.
```
Comment (D4nte) [inline 0013-secure-http-api.adoc]:
```suggestion
TLS requires a certificate (private/public keypair) to be present on the server i.e., cnd.
```

File: 0013-secure-http-api.adoc [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+There are two aspects to encrypting traffic between cnd to the client.
+1. The HTTP API exposed by cnd needs to be encrypted using TLS.
+2. *Our* client implementation, `comit-i` needs to be served by cnd using TLS.
+
+=== cnd
+
+Adding TLS to cnd should be straightforward.
+
+There are two parts to this:
+1. Exposing the HTTP API vial TLS (so clients can connect to cnd using TLS)
```
Comment (D4nte) [inline 0013-secure-http-api.adoc:120]:
```suggestion
1. Exposing the HTTP API over TLS (so clients can connect to cnd using TLS)
```

File: 0013-secure-http-api.adoc [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+2. Serving `comit-i` via TLS (so the browser can use `https://` addresses)
```
Comment (D4nte) [inline 0013-secure-http-api.adoc:121]:
```suggestion
2. Serving `comit-i` over TLS (so the browser can use `https://` addresses)
```

File: 0013-secure-http-api.adoc [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+Since both of these are done by calls to Warp they can both be done in the same manner.
+TLS is supported by Warp (the web server framework we use).
+Requires adding a field in the config file and command line options e.g., `--tls-key=/path/to/key` and `--tls-cert=/path/to/cert`.
+Adds an extra initial setup step for the user of cnd to ensure valid certificate is available on the host machine.
+
+=== comit-i
+
+`comit-i` must consume the HTTP API exposed by cnd using TLS
+I was unable to find any information on the internet on how to do this.
```
Comment (D4nte) [inline 0013-secure-http-api.adoc]: I assume you just update the cnd REST API end point from `http` to `https`.

File: 0013-secure-http-api.adoc [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+The cnd stuff is straightforward.
+
+.Recommendations are as follows:
+* Implement TLS for cnd using Warp's TLS support.
+** Enable TLS by default, add option `--no-tls` to disable it.
+** Look for keypair in well known location, configurable via configuration file and command line options.
+* Leave generation of valid certificate up to the user of cnd but add documentation/hints on how to do this and what options are available (based on information in this spike).
+
+As for `comit-i` my findings were less positive.
+Recommend splitting `comit-i` up into separate parts for the front end and back end and re-writing the back end in Rust.
```
Comment (D4nte) [inline 0013-secure-http-api.adoc]: I don't understand why you would need that.

File: 0013-secure-http-api.adoc [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+Use case: connecting to `comit-i` with the machine running cnd.
+Create a self signed certificate.
+
+.pros
+* Free
+* Within the realm of typical developer knowledge
+
+.cons
+* Adds complexity to the set up of cnd
```
Comment (D4nte) [inline 0013-secure-http-api.adoc:100]: Why not:
- cnd generates certificates when starting up if not yet existing and if `webgui` is activated and `webgui.tls` (new option) is activated. This will be a CA, Public Cert and Private Cert
- User is told to install the freshly generated CA

That's it.

Feel free to at `lnd` as it does generate certificates for its gRPC API on the spot as described above (it does not generate a CA though, only a `tls.cert` and `tls.key`).

File: 0013-secure-http-api.adoc [hunk 10]
```diff
@@ -52,34 +52,98 @@ The industry standard for encrypting traffic over a network is Transport Layer S
...
+The cnd stuff is straightforward.
+
 .Recommendations are as follows:
 * Implement TLS for cnd using Warp's TLS support.
 ** Enable TLS by default, add option `--no-tls` to disable it.
 ** Look for keypair in well known location, configurable via configuration file and command line options.
+* Leave generation of valid certificate up to the user of cnd but add documentation/hints on how to do this and what options are available (based on information in this spike).
+
+As for `comit-i` my findings were less positive.
+Recommend splitting `comit-i` up into separate parts for the front end and back end and re-writing the back end in Rust.
```
Comment (thomaseizinger) [inline 0013-secure-http-api.adoc]: Is your idea that we can then serve `comit-i` on `http` again?
I am afraid, this does not really help us :(
One of the reasons why it would be useful to serve `comit-i` via HTTPS is to make integrations with a Ledger Nano S possible (the browser only allows you to talk to a USB device on HTTPS)

**Review by thomaseizinger — APPROVED** [review]
I am starting to like this :)

It looks like our best option is to on-demand create a certificate for cnd if the user wants to have TLS for the API.

File: 0013-secure-http-api.adoc (hunk 7 above)
Comment (tcharding) [reply 0013-secure-http-api.adoc]: Unless I'm misunderstanding the architechture, I think it is more complicated than that.  `comit-i` consumes the REST API via typescript code, not via the web browser so `comit-i` has to do the TLS handshake.

File: 0013-secure-http-api.adoc (hunk 8 above)
Comment (tcharding) [reply 0013-secure-http-api.adoc]: Because of my comment above, if it is correct.

File: 0013-secure-http-api.adoc (hunk 9 above)
Comment (tcharding) [reply 0013-secure-http-api.adoc:100]: I think I get what you are saying but am confused by the terminology - I'm only just across all the TLS certificate terms as is.  Either way we still either make the user click through security warnings or require root to install the CA.

File: 0013-secure-http-api.adoc (hunk 10 above)
Comment (tcharding) [reply 0013-secure-http-api.adoc]: The idea is to serve `comit-i` using HTTPS.

File: 0013-secure-http-api.adoc (hunk 9 above)
Comment (tcharding) [reply 0013-secure-http-api.adoc:100]: I do not think that `cnd` should create the certificate.  I think we should stick to the UNIX principle of doing one thing.  The user can generate the certificate.

**Comment by tcharding** [comment]
> It looks like our best option is to on-demand create a certificate for cnd if the user wants to have TLS for the API.

As commented above, I don't think we should be creating the certificate in `cnd`, I feel that we should leave it up to the user to do so.

**Review by thomaseizinger — APPROVED** [review]
Thanks for adding the final notes @luckysori !

FYI @tcharding, the final notes are the conclusion of our discussion this morning :)

**Review by bonomat — APPROVED** [review]
